easy_logging.go

```diff
@@ -117,6 +129,9 @@ func createLogWriter(logPath string) (io.Writer, *os.File, error) {
 }
 
 func allowedToInitialize(clientConfigFileInput string) bool {
+	easyLoggingInitTrials.mu.RLock()
```

Comment:
maybe let's lock only in initEasyLogging and reset, and remove locking from allowedToInitialize, setInitTrial, increaseReconfigureCounter. Otherwise many initialisations at the same time are still possible.

Reply:
Hmm, I don't see how that solves the issue if we're allowed to write to it in any of `setInitTrial` or `increaseReconfigureCounter` and read from it in `allowedToInitialize`. Since the variable is global it can be read or written at the same time at any point.

If I remove locking from `allowedToInitialize`, `setInitTrial` and `increaseReconfigureCounter` I get a race in these methods. Running the reproducible example in the ticket:

```text
› go test . -race 
==================
WARNING: DATA RACE
Write at 0x0001021668c0 by goroutine 20:
  github.com/snowflakedb/gosnowflake.(*initTrials).setInitTrial()
[SNIP]
Previous read at 0x0001021668c0 by goroutine 21:
  github.com/snowflakedb/gosnowflake.allowedToInitialize()

==================
WARNING: DATA RACE
Write at 0x0001021668c8 by goroutine 26:
  github.com/snowflakedb/gosnowflake.(*initTrials).setInitTrial()
[SNIP]
Previous write at 0x0001021668c8 by goroutine 19:
  github.com/snowflakedb/gosnowflake.(*initTrials).setInitTrial()
```

I opted to use a `RWLock` so we can have multiple reads in `allowedToInitialize` at the same time and avoid unnecessary locking there.

Reply:
Ah, never mind, now I see what you mean, you want me to take a lock through the whole `initEasyLogging`? I can do that, but that would introduce a new race if someone calls `setInitTrial`, `increaseReconfigureCounter` or `allowedToInitialize` outside of those methods.

It's also a bit weird to take the lock far away from where we do the write in my opinion but maybe that's just personal preference. Also this means that I'll have to take a full lock (read and write) even when a short read lock would be sufficient in `allowedToInitialize`.

Reply:
Pushed a change with your suggested changes since where to lock is up to you. A bigger rewrite to leverage something like a `sync.Once` should probably be made in a separate PR and not as a part of this bug fix.